core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java

```diff
@@ -94,110 +97,155 @@ public PrimaryShardAllocator(Settings settings) {
         logger.debug("using initial_shards [{}]", NODE_INITIAL_SHARDS_SETTING.get(settings));
     }
 
-    public void allocateUnassigned(RoutingAllocation allocation) {
-        final RoutingNodes routingNodes = allocation.routingNodes();
-        final MetaData metaData = allocation.metaData();
-
-        final RoutingNodes.UnassignedShards.UnassignedIterator unassignedIterator = routingNodes.unassigned().iterator();
-        while (unassignedIterator.hasNext()) {
-            final ShardRouting shard = unassignedIterator.next();
-
-            if (shard.primary() == false) {
-                continue;
-            }
+    /**
+     * Is the allocator responsible for allocating the given {@link ShardRouting}?
+     */
+    private static boolean isResponsibleFor(final ShardRouting shard) {
+        return shard.primary() // must be primary
+                && shard.unassigned() // must be unassigned
+                // only handle either an existing store or a snapshot recovery
+                && (shard.recoverySource().getType() == RecoverySource.Type.EXISTING_STORE
+                        || shard.recoverySource().getType() == RecoverySource.Type.SNAPSHOT);
```

Comment:
I like that this is explicit in the file, I think it makes reading the code easier